✨ feat(lock): add SQLite-based ReadWriteLock

leventov opened the pull request:
The existing lock classes (`FileLock`, `SoftFileLock`) only support exclusive locking -- every caller blocks every other caller, even when most only need to read. This is fine for write-heavy workloads, but read-heavy ones pay an unnecessary serialization cost. 🔒

This PR adds `ReadWriteLock`, a cross-process shared/exclusive lock backed by SQLite transactions. Multiple processes can hold a read lock simultaneously while write locks remain exclusive. The lock is reentrant within the same mode (nested reads or nested writes from the same thread), but upgrading from read to write or downgrading from write to read raises `RuntimeError` to prevent deadlocks. Write locks are pinned to the acquiring thread.

The implementation uses SQLite's built-in `BEGIN TRANSACTION` (shared) and `BEGIN EXCLUSIVE TRANSACTION` (exclusive) with `journal_mode=MEMORY` rather than WAL, because WAL would allow readers to proceed during an exclusive write -- exactly the behavior we need to prevent. A double-check locking pattern serializes concurrent threads through a `threading.Lock` before touching SQLite, and `busy_timeout` is recomputed after each potentially-blocking pragma to keep timeout accounting accurate. Instances are singletons by default (keyed by resolved path via a `WeakValueDictionary`), matching the pattern established by `BaseFileLock`'s `is_singleton` parameter.

📝 The docs (`docs/index.rst`) have been substantially reworked: a new section hierarchy (`Usage` / `Lock types`), platform resolution details for `FileLock`, a dedicated `ReadWriteLock` section with usage examples, limitations for every lock type, async variant mentions in "Choosing the right lock", and singleton lock documentation. The changelog has been backfilled from v3.12.1 through v3.20.3. Code examples now use `pathlib.Path` and descriptive variable names throughout. Existing docstrings in `_api.py`, `asyncio.py`, and `_soft.py` have been expanded with `:param:` / `:raises:` / `:return:` tags and `versionadded` directives.

♻️ The `_read_write.py` source extracts shared acquire logic into `_acquire()`, `_validate_reentrant()`, and `_configure_and_begin()` to eliminate ~50 lines of duplication between `acquire_read` and `acquire_write`. Internal attributes use underscore-prefixed names (`_con`, `_lock_level`, etc.) and imports are relative throughout.

A PyPy-specific `test_virtualenv` skip was added due to a segfault in PyPy's `_sqlite3.Cursor.__del__._reset` during garbage collection -- no upstream issue was found for this bug.

leventov committed:
SQLite-based read-write lock

gaborbernat commented on src/filelock/_read_write.py:
```diff
@@ -0,0 +1,80 @@
+        super().__init__(lock_file, timeout, blocking)
+        self.procLock = threading.Lock()
+        self.con = sqlite3.connect(self._context.lock_file, check_same_thread=False)
+        # Redundant unless there are "rogue" processes that open the db
```
Please align comments to 120 charachters 🙇

gaborbernat commented on src/filelock/_read_write.py:
```diff
@@ -0,0 +1,80 @@
+
+
+class _SQLiteLock(BaseFileLock):
+    def __init__(self, lock_file: str | os.PathLike[str], timeout: float = -1, blocking: bool = True) -> None:
```
Prefer 
```
from os import PathLike
```

leventov committed:
Refactor, fix bugs

leventov committed:
Reduce timeout from _transaction_lock.acquire() wait

gaborbernat reviewed (changes requested):
Looks better, can you add tests please?

leventov committed:
Add tests, move journal_mode=MEMORY pragma into transaction blocks

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

gaborbernat reviewed (changes requested):
CI failing.

leventov committed:
Prohibit re-entering write lock from concurrent threads

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

gaborbernat committed:
🧪 test(read_write): add unit tests for 100% coverage

The existing test suite relied entirely on multiprocessing, which coverage
cannot track across process boundaries. This left _read_write.py at 47%
measured coverage despite the code being well-exercised.

Added in-process unit tests covering all branches: timeout_for_sqlite,
singleton metaclass, acquire/release, reentrance, mode conflicts,
OperationalError handling, double-check race paths, and context managers.
Parametrized read/write symmetric tests to reduce duplication. Fixed ruff
violations (FBT, UP031, B904, ANN) and removed dead duplicate test helpers.

Signed-off-by: Bernát Gábor <bgabor8@bloomberg.net>

gaborbernat committed:
📝 docs: add docstrings, backfill changelog

The project had inconsistent docstrings (terse one-liners, malformed
RST tags, typos) and the changelog stopped at v3.12.0 despite 20+
releases since. The new ReadWriteLock class also shipped without any
docstrings or user-facing documentation.

Expand class/method docstrings across _api.py, _soft.py, asyncio.py,
and _read_write.py with RST-style :param:/:raises:/:return: tags and
versionadded annotations. Fix ::return: -> :return: in asyncio
properties and "in every case/" typo in release() docstrings.

Rewrite docs/index.rst: replace "FileLock vs SoftFileLock" with
"Choosing the right lock", add ReadWriteLock usage section, document
asyncio support and singleton locks. Backfill changelog from v3.12.1
through v3.20.3.

Signed-off-by: Bernát Gábor <bgabor8@bloomberg.net>

gaborbernat committed:
♻️ refactor(read_write): extract shared acquire logic

acquire_read and acquire_write had ~50 lines of duplicated structure
(reentrant check, double-check after transaction lock, SQLite setup,
error handling). Extract into _acquire, _validate_reentrant, and
_configure_and_begin to eliminate the duplication.

Also: rename con to _con (internal detail), fix absolute import to
relative, widen AcquireReturnProxy type to accept ReadWriteLock, merge
duplicate test helpers, improve chain_reader param naming, add comments
to bare time.sleep calls, fix singleton cache fixture teardown, remove
dead __main__ block, skip test_virtualenv on PyPy due to SQLite
Cursor.__del__ segfault.

gaborbernat committed:
💄 style: replace single-letter variables with descriptive names

Single-letter process/thread variables (p1, p2, t, p, e) made tests
harder to scan. Renamed to role-based names: reader, writer, holder,
contender, successor, crashing, worker, thread, etc.

gaborbernat committed:
💄 style: rewrap docstrings to fill 120-char lines

Docstring lines were wrapping at ~80 chars leaving significant unused
space. Reflowed all multi-line docstrings to fill closer to the 120-char
limit, reducing total line count. Also removed trailing backslash
continuations in favor of natural line breaks.

gaborbernat committed:
🐛 fix(read_write): resolve ty type checker errors

The `ty` type checker flagged four errors: `cls.get_lock` was unresolvable
inside the metaclass `__call__`, the 2-arg `super()` call was invalid for
metaclass contexts, and `Timeout()` received `PathLike` instead of `str`.

Moved singleton resolution logic into the metaclass `__call__` directly
(matching the `FileLockMeta` pattern in `_api.py`), making `get_lock` a
thin wrapper. Added `os.fspath()` conversion in `__init__` so `lock_file`
is always `str` downstream.

gaborbernat committed:
👷 ci(release): replace tag-triggered with manual workflow

The tag-triggered release pipeline required creating a tag manually
before releasing, with no changelog generation or GitHub release notes.

Switched to a `workflow_dispatch` pipeline (modeled after toml-fmt)
where maintainers choose the semver bump level from the Actions UI.
The workflow computes the next version, auto-generates changelog from
git history and PR metadata, builds the package with the correct
version tag for hatch-vcs, publishes to PyPI, and creates a GitHub
release with generated notes. Removed the handwritten v3.21.0
unreleased changelog entry since it will be auto-generated at release
time.

Signed-off-by: Bernát Gábor <bgabor8@bloomberg.net>

gaborbernat committed:
🐛 fix(read_write): prevent PyPy segfault in SQLite Cursor.__del__

PyPy's GC collects ReadWriteLock objects during fixture teardown,
triggering Cursor.__del__._reset on already-finalized SQLite state.

Explicitly close connections in the test fixture before clearing the
singleton cache so GC has nothing dangerous to finalize. Guard
__del__ with contextlib.suppress for defense in depth.

Signed-off-by: Bernát Gábor <bgabor8@bloomberg.net>

gaborbernat committed:
🐛 fix(read_write): replace __del__ with atexit to prevent PyPy segfault

PyPy's GC can run sqlite3.Cursor.__del__._reset at any allocation point
after a connection is closed, causing a segfault. This happens because
__del__ on ReadWriteLock would call rollback() (creating cursors) then
close(), leaving orphan cursors for later GC to trip over.

Replace __del__ with an atexit handler that closes all connections
before interpreter finalization. Add an explicit close() method for
deterministic cleanup.
